### PR TITLE
Add guild for developing modules and functionals

### PR DESCRIPTION
**Pull request by shendiaomo**

### CODE REVIEW

File: doc/developer_guild.md [hunk 1]
```diff
@@ -0,0 +1,471 @@
+# Extend Functionalities of GoTorch
+
+At the very core of PyTorch is *libtorch*, which is a C++ library that contains
+three levels of abstraction:
+
+1. *Native function*s are low-level functions, most of which operate tensors.
+   They are the basic mathematical build blocks of deep neural networks.
+1. *Functional*s are high-level functions built on native functions, each
+   functional may call several native functions.
+1. *Module*s are c++ classes/structs/templates that encapsulate multiple native
```
Comment (wangkuiyi) [inline doc/developer_guild.md:10]: I thought libtorch doesn't contain modules, which are Python classes.

File: doc/developer_guild.md [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+   functions or functionals.
+
+Modules and functionals form the Pytorch C++ frontend.  AI practitioners can
```
Comment (wangkuiyi) [inline doc/developer_guild.md:13]: Pytorch => PyTorch

Comment (wangkuiyi) [inline doc/developer_guild.md:13]: According to https://pytorch.org/cppdocs/frontend.html, PyTorch C++ frontends includes

- torch::Tensor, Automatically differentiable, efficient CPU and GPU enabled tensors
- torch::nn, A collection of composable modules for neural network modeling
- torch::optim, Optimization algorithms like SGD, Adam or RMSprop to train your models
- torch::data, Datasets, data pipelines and multi-threaded, asynchronous data loader
- torch::serialize, A serialization API for storing and loading model checkpoints
- torch::python, Glue to bind your C++ models into Python
- torch::jit, Pure C++ access to the TorchScript JIT compiler

These include but not limited to modules or functionals.

I don't think it is necessary to mention the concept of PyTorch C++ frontend.

File: doc/developer_guild.md [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+directly use libtorch modules or functionals to build their AI applications.
+
+libtorch contains about 1600 native functions, they are the foundations of
```
Comment (wangkuiyi) [inline doc/developer_guild.md:16]: This paragraph highly duplicate with the `wrap_native_functions.md`. I would recommend to remove it.

File: doc/developer_guild.md [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+in package `torch.nn.functional` and modules classes in `torch.nn` call native
+functions via [`pybind`](https://github.com/pybind/pybind11).  In GoTorch, the
+corresponding packages `gotorch/nn/functional` and `gotorch/nn` call Go wrappers
+of native functions.
+
+The tutorial demonstrates how to wrap libtorch native functions or functionals
+into GoTorch using [Cgo](https://blog.golang.org/cgo), and how to build modules
+based on native functions and functionals.
+
+## Wrap PyTorch Native Functions
```
Comment (wangkuiyi) [inline doc/developer_guild.md:27]: We already have `wrap_native_functions.md`!

File: doc/developer_guild.md [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+   Because `C.Tensor` is a pointer to `at::Tensor`, the line `*result = new
+   at::Tensor(c)` in the C wrapper makes `t` pointing to the newly allocated
+   tensor on the heap.
+1. It calls `MustNil` to check the C string returned from the C wrapper.
+   `MustNil` panics if the string is not `nullptr`.
+1. It calls `gotorch.SetTensorFinalizer` to attach the finalizer to the tensor
+   returned by `C.MM`.
+1. It returns a value of the Go type `Tensor` that encapsulates `t`.
+
+## Define GoTorch Functionals
```
Comment (wangkuiyi) [inline doc/developer_guild.md:227]: Let us do not remove the existing file. But add a new one --  Porting PyTorch Functions and Modules

File: doc/developer_guild.md (hunk 1 above)
Comment (shendiaomo) [reply doc/developer_guild.md]: Done.

File: doc/developer_guild.md (hunk 2 above)
Comment (shendiaomo) [reply doc/developer_guild.md]: Done.

Comment (shendiaomo) [reply doc/developer_guild.md]: Done.

File: doc/developer_guild.md (hunk 3 above)
Comment (shendiaomo) [reply doc/developer_guild.md]: Done.

File: doc/developer_guild.md (hunk 4 above)
Comment (shendiaomo) [reply doc/developer_guild.md]: Done.

File: doc/developer_guild.md (hunk 5 above)
Comment (shendiaomo) [reply doc/developer_guild.md]: Done.

**Pull request merged**